### PR TITLE
Adding Terraform Plugin Protocol top-level section (#241)

### PR DESCRIPTION
**Pull request by bendbennett**
### What

Split `Terraform Plugin Protocol` from the `How Terraform Works` page into a separate top-level page.

### Why

This change supplies Terraform provider developers with additional information about the RPCs that are issued in response to Terraform commands and the methods within the Terraform plugin framework that are called as a consequence. This provides further clarity on which RPCs call which Terraform plugin framework methods and the sequence the RPCs are sent during the execution of a specific Terraform command

Closes: #241 

----------

### Merge Checklist
_If items do not apply to your changes, add (N/A) and mark them as complete._

#### Pull Request
- [x] One or more labels describe the type of change (e.g. clarification) and associated product (e.g. tfc).
- [x] Description links to related pull requests or issues, if any.

#### Content
- **N/A** Redirects have been added for moved, renamed, or deleted pages. This requires a separate PR in the [`terraform-website` repository](https://github.com/hashicorp/terraform-website) `redirects.next.js` file.
- **N/A** API documentation and the API Changelog have been updated. 
- **N/A** Links to related content where appropriate (e.g., API endpoints, permissions, etc.).
- **N/A** Pages with related content are updated and link to this content when appropriate.
- [x] Sidebar navigation files have been updated for added, deleted, reordered, or renamed pages.
- [x] New pages have metadata (page name and description) at the top.
- [ ] New images are 2048 px wide. They have HashiCorp standard annotation color (#F92672) and format (rectangle with rounded corners), blurred sensitive details (e.g. credentials, usernames, user icons), and descriptive alt text in the markdown for accessibility.
- **N/A** New code blocks have the correct syntax and line breaks to eliminate horizontal scroll bars.
- **N/A** UI elements (button names, page names, etc.) are bolded.
- [x] The Vercel website preview successfully deployed.

#### Reviews
- [x] I or someone else reviewed the content for technical accuracy.
- [x] I or someone else reviewed the content for typos, punctuation, spelling, and grammar.

### CODE REVIEW

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:46]: It might be helpful if this table gave a brief description of the RPC in a new column. 👍 e.g. `GetProviderSchema` Returns schema information for the entire provider, including the provider schema, provider meta schema, all data source schemas, and all resource schemas.

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:77]: The duplicate calls to the RPC are an implementation detail of how Terraform currently works and not part of any contract with providers. I'd suggest removing duplicates here and below.

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:82]: I believe we should leave framework specific information onto its own framework documentation, as this top-level documentation section is serving terraform-plugin-go, terraform-plugin-sdk, terraform-plugin-framework, and other generic gRPC implementations.

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:74]: Terraform implementation details will dictate the actual order and are not protected by any sort of contracts with providers. The ordering is determined by the DAG and any parallelism. In fact, we may be want to be even more generic and just mention that here's the _potential_ RPCs. If there are no data sources in the configuration, then `ValidateDataResourceConfig` is not called, for example.

```suggestion
The general sequence of RPCs is as follows:
```

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:144]: We may want to confer with the core team to determine if they want to document this level of detail. For provider developers, I believe the important details are which RPCs are potentially called.

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:165]: This seems like it should be its own framework documentation page since its SDK specific and could change over time.

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bendbennett) [reply website/docs/plugin/terraform-plugin-protocol.mdx:144]: I'll remove this for now.

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:57]: Terraform allows for plans without configuration changes (e.g. only updating `Computed`-only values). This is so providers can signal potential provider behavior changes via a plan-apply workflow. I believe the below wording should capture what we are going for here.

```suggestion
| `ApplyResourceChange`        | `ApplyResourceChange`        | Called when a practitioner has approved a planned change. The provider is to apply the changes contained in the plan, and return a resulting state matching the given plan. |
```

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:113]:
```suggestion
  - only called if resource already exists in state
```

File: website/docs/plugin/how-terraform-works.mdx
Comment (bflad) [inline website/docs/plugin/how-terraform-works.mdx:151]: I think these need to stay on this page (or be changed to inline links instead of detached links)

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:120]: These detached links are not present on this new page.

```suggestion
```

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bflad) [inline website/docs/plugin/terraform-plugin-protocol.mdx:85]: This diagram is missing `ReadResource`

**Review by bflad — APPROVED** [review]
Overall looks good to me 🚀  Just some things to consider below before merging.

File: website/docs/plugin/terraform-plugin-protocol.mdx
Comment (bendbennett) [reply website/docs/plugin/terraform-plugin-protocol.mdx:85]: Well spotted! Fixed.

**Review by bflad — APPROVED** [review]

**Comment by bflad** [comment]
Pulling this in to coordinate publishing with #252, which has links pointing to the new page.

**Pull request merged**
